### ISSUE TITLE
Focus is lost after closing accessible view

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#188822]
Testing #188537

- Focus on a hover
- Press alt+f2
- Press escape

It seems like focus just goes to the 'body' element and I have to tab to get focus back to where I was before. I'd expect it to preserve focus in the same context, same as the command palette.

**Comment by meganrogge** [comment]
To properly fix this, we would need to trap the element that's focused before the hover is shown - since by the time the accessible view code is reached, the focus element is the hover itself and there's no way currently to get that from the context view service (for non-editor hovers)

In the meantime, i'm going to see if we can focus the hover - at least from there, escaping will return the focus to the correct element

**Comment by meganrogge** [comment]
@Tyriar without a big refactor, can you think of a way to get the element that was focused before the hover opened?

**Comment by roblourens** [comment]
Can you just grab `document.activeElement` before showing the accessible view?

**Comment by meganrogge** [comment]
that's what I tried and it didn't work because, as mentioned above, the active element is the hover

**Comment by roblourens** [comment]
Right but I thought that you would want to restore focus to the hover

**Comment by meganrogge** [comment]
That doesn't work because that element is hidden (it's a context view) on blur when the accessible buffer is opened. So I can call `focus` on it, but it's not visible and even if it were, wouldn't be in the right spot. we'd have to somehow save the args passed into `showContextView` and invoke that. seems really messy, thus need to defer and rethink

**Comment by roblourens** [comment]
It doesn't get hidden for me. But yeah maybe that's not a guarantee

**Comment by Tyriar** [comment]
The hover goes away for me, the best thing would be to prevent the hover from hiding until the a11y view is gone imo. We could fix this by adding `lockHover` and `unlockHover` which would prevent it from hiding, then using them in `HoverAccessibleViewContribution`. Make sure the hover is focused before calling `unlockHover`.

**Comment by meganrogge** [comment]
The hover is getting disposed of before it is hidden - on mouse out

**Comment by meganrogge** [comment]
Yeah it's coming from a context view call of `hide` - 👀 

![Screenshot 2023-07-27 at 9 20 50 AM](https://github.com/microsoft/vscode/assets/29464607/87fbf289-511e-4957-85b5-de642e417a5c)

**Comment by Tyriar** [comment]
@meganrogge that would come from the hover or hover service, `lockHover` needs to prevent that

**Comment by meganrogge** [comment]
oh - I think it's because we use the context view for the accessible view - so when we call show of the accessible view, the current context view is disposed of (the hover) 

![Screenshot 2023-07-27 at 9 22 27 AM](https://github.com/microsoft/vscode/assets/29464607/30d1f7d8-f552-49e4-83e5-9d4a5dd3e8ca)

**Comment by meganrogge** [comment]
yeah i tried to do that like this
![Screenshot 2023-07-27 at 9 23 20 AM](https://github.com/microsoft/vscode/assets/29464607/d8fc1b6e-1053-4956-8e33-8baba58ff964)

**Comment by meganrogge** [comment]
I think we will have to go this route for now
https://github.com/microsoft/vscode/assets/29464607/ccc8610d-7fac-420f-98d5-1d14023a0766

**Comment by meganrogge** [comment]
Verify that when you focus a hover in the editor (via cmd/ctrl + k + i) and in the GHPR panel
- use alt + f2 to open the accessible view
- press escape and the hover should be focused
- press escape again and the previously focused item should have focus

**Comment by roblourens** [comment]
![Recording 2023-07-27 at 09 49 24](https://github.com/microsoft/vscode/assets/323878/15467352-6eb2-4171-89eb-31e91d1a08fc)

I still wonder why this already works fine for me lol

**Comment by meganrogge** [comment]
you're in the editor @roblourens where I fixed this issue yesterday

**Comment by Tyriar** [comment]
@roblourens the editor hover is a monaco content widget, as opposed to a context view

**Comment by alexr00** [comment]
This works great in the editor, but in the tree view I see the following:
- `ctrl+k+i` doesn't focus the hover
- When I `esc` from the accessible view, the hover gets focus. When I then `esc` from the hover, the previously focused element does not get focus.

**Comment by Tyriar** [comment]
@meganrogge we don't handle `_lastFocusedElementBeforeOpen` correctly in this case:

https://github.com/microsoft/vscode/blob/9b81a0d764c5ff811562792e1b23eb583b0042e3/src/vs/workbench/services/hover/browser/hoverService.ts#L28

**Comment by meganrogge** [comment]
Oh, this is happening partly because `options.trapFocus` is not set for the initial `showHover` call. I'm not sure why that would be the case. So `_lastFocusedElementBeforeOpen` is undefined. 

If that were fixed, then it would just be a matter of passing in `skipLastFocusedUpdate` to `showHover` from `showAndFocusLastHover` such that it doesn't override the element that was set on the initial `show` call.

https://github.com/microsoft/vscode/assets/29464607/3ebe9521-4dbd-430a-be3f-55566c377f60

**Comment by meganrogge** [comment]
I think we always want to `trapFocus` for screen reader users, correct @Tyriar ? not sure why this hasn't been reported as bug before.
